### ISSUE TITLE
SSDP binds to INADDR_ANY

### ISSUE THREAD

**Issue by whyman** [pupnp/pupnp#227]
There is a lot of logic to find suitable IPv4/v6 addresses, but by the looks of it that is ignored by SSDP server code which just binds with `INADDR_ANY` regardless?

https://github.com/pupnp/pupnp/blob/643f9c3ea21b0312865e9e8ed49653850cbd3a55/upnp/src/ssdp/ssdp_server.c#L973-L977

Similar code exists for IPv6 too.

I am not a sockets programming expert, but looks odd to me?

**Comment by mrjimenez** [comment]
I think that the code that calls the function you show will only fail if an IPV4 interface was originally found:

https://github.com/pupnp/pupnp/blob/c400c1e01b53190a8f042532ec8f379850813064/upnp/src/ssdp/ssdp_server.c#L1493-L1504

Since you will already need a SSDP socket on IPV4, it seems reasonable to fail, but maybe we could insist and try IPV6 if it has been selected and run in IPV6 only mode? I agree that the library initialization is a bit confusing.

**Comment by mrjimenez** [comment]
Hum, after reading it again, I think I did not get your point in the first place.

I think you are right, it blindly binds to `INADDR_ANY`. Maybe the right thing is to bind to the same addresses it found before.

Wasn't that your point? :confused:

**Comment by whyman** [comment]
Yes, that was my point 😅

**Comment by mrjimenez** [comment]
Hi @whyman — confirmed.

The three SSDP server-socket creation functions (`create_ssdp_sock_v4`, `create_ssdp_sock_v6`, `create_ssdp_sock_v6_ula_gua`) all bind to a wildcard address (`INADDR_ANY` / `in6addr_any`) even though the specific interface address (`gIF_IPV4` / `gIF_IPV6` / `gIF_IPV6_ULA_GUA`) is already available and is already used immediately after the bind for multicast group membership (`IP_ADD_MEMBERSHIP`, `IPV6_JOIN_GROUP`) and multicast interface (`IP_MULTICAST_IF`).

The result is that, on a multi-homed host, the SSDP socket accepts UDP traffic arriving on interfaces other than the one selected at `UpnpInit2()` time.

The fix is to bind to the specific interface address at all three sites. On Linux (and most POSIX stacks) receiving multicast traffic is still correctly gated by the `IP_ADD_MEMBERSHIP` join — binding to a unicast address does not suppress delivery of multicast datagrams the socket has joined.

**Comment by mrjimenez** [comment]
Draft PR opened: https://github.com/pupnp/pupnp/pull/579

Branch `fix/issue-227-ssdp-binds-to-inaddrany` has been created with the POC regression tests already committed. The tests are expected to fail until the fix is implemented.

**Comment by whyman** [comment]
Woop thanks @mrjimenez / Claude!

**Comment by mrjimenez** [comment]
**Update — regression in PR #579**

The fix in PR #579 introduced a regression: binding the SSDP sockets to the specific interface unicast address prevents the Linux kernel from delivering multicast M-SEARCH packets to them, breaking device discovery entirely (reported by @KarlStraussberger / Gerbera).

The underlying issue is that `IP_ADD_MEMBERSHIP` with `imr_interface = gIF_IPV4` already scopes multicast group membership to the correct interface. Changing the `bind()` address from `INADDR_ANY` to the unicast address is not needed for interface scoping and breaks multicast reception.

PR #601 reverts the bind sites and replaces the flawed test with a behavioral one that sends a real M-SEARCH and verifies the library responds. This will be released as 2.0.1.
